cpp/src/strings/json/json_path.cu

```diff
@@ -0,0 +1,917 @@
+    if (src_size > 0) {
+      char* dst             = out_buf ? out_buf + output_offsets[tid] : nullptr;
+      size_t const dst_size = out_buf ? output_offsets[tid + 1] - output_offsets[tid] : 0;
+
+      parse_result result;
+      json_output out;
+      thrust::tie(result, out) =
+        get_json_object_single(chars + offsets[tid], src_size, commands, dst, dst_size);
+      output_size = out.output_len;
+      if (out.element_count > 0 && result == parse_result::SUCCESS) { is_valid = true; }
+    }
+
+    // filled in only during the precompute step
+    if (!out_buf) { output_offsets[tid] = static_cast<size_type>(output_size); }
+  }
+
+  // validity filled in only during the output step
+  if (out_validity) {
+    uint32_t mask = __ballot_sync(0xffffffff, is_valid);
+    // 0th lane of the warp writes the validity
+    if (!(tid % cudf::detail::warp_size) && tid < num_rows) {
+      out_validity[cudf::word_index(tid)] = mask;
+    }
+  }
+}
+
+/**
+ * @copydoc cudf::strings::detail::get_json_object
+ */
+std::unique_ptr<cudf::column> get_json_object(cudf::strings_column_view const& col,
+                                              cudf::string_scalar const& json_path,
+                                              rmm::cuda_stream_view stream,
+                                              rmm::mr::device_memory_resource* mr)
+{
+  // preprocess the json_path into a command buffer
+  std::tuple<rmm::device_uvector<path_operator>, int, bool> preprocess =
+    build_command_buffer(json_path, stream);
+  CUDF_EXPECTS(std::get<1>(preprocess) <= max_command_stack_depth,
+               "Encountered json_path string that is too complex");
+
+  // allocate output offsets buffer.
+  auto offsets = cudf::make_fixed_width_column(
+    data_type{type_id::INT32}, col.size() + 1, mask_state::UNALLOCATED, stream, mr);
+  cudf::mutable_column_view offsets_view(*offsets);
+
+  // if the query is empty, return a string column containing all nulls
+  if (std::get<2>(preprocess)) {
+    thrust::generate(rmm::exec_policy(stream),
+                     offsets_view.head<size_type>(),
+                     offsets_view.head<size_type>() + offsets_view.size(),
+                     [] __device__() { return 0; });
+    return cudf::make_strings_column(
+      col.size(),
+      std::move(offsets),
+      cudf::make_fixed_width_column(
+        data_type{type_id::INT8}, 0, mask_state::UNALLOCATED, stream, mr),
+      col.size(),
+      cudf::detail::create_null_mask(col.size(), mask_state::ALL_NULL, stream, mr),
+      stream,
+      mr);
+  }
+
+  cudf::detail::grid_1d const grid{col.size(), 512};
+
+  // preprocess sizes (returned in the offsets buffer)
+  get_json_object_kernel<<<grid.num_blocks, grid.num_threads_per_block, 0, stream.value()>>>(
+    col.chars().head<char>(),
+    col.offsets().head<size_type>(),
+    std::get<0>(preprocess).data(),
+    offsets_view.head<size_type>(),
+    nullptr,
+    nullptr,
+    col.size());
```

Comment:
This is processing a string per thread.
Why not use a `column_device_view` and then simply use `element<string_view>()` rather than accessing the strings columns' child columns data directly? This saves all the hassle of offseting offsets, computing the size by diff-ing offsets, etc.